src/compression.c

```diff
@@ -0,0 +1,154 @@
+    // so we link it statically and set functions pointers directly
+    zlib_ver = zlibVersion;
+    zlib_flags = zlibCompileFlags;
+    inflateInit_f = inflateInit_;
+    inflateInit2_f = inflateInit2_;
+    inflateEnd_f = inflateEnd;
+    inflate_f = inflate;
+    zError_f = zError;
+#else
+#define CHECK_DL(op) do{ \
+if ((op) != 0)           \
+goto on_error;           \
+} while(0)
+
+    CHECK_DL(uv_dlopen(SO_lib(libz), &zlib));
+    CHECK_DL(uv_dlsym(&zlib, "zlibVersion", (void **) &zlib_ver));
+    CHECK_DL(uv_dlsym(&zlib, "zlibCompileFlags", (void **) &zlib_flags));
+    CHECK_DL(uv_dlsym(&zlib, "inflateEnd", (void **) &inflateEnd_f));
+    CHECK_DL(uv_dlsym(&zlib, "inflateInit_", (void **) &inflateInit_f));
+    CHECK_DL(uv_dlsym(&zlib, "inflateInit2_", (void **) &inflateInit2_f));
+    CHECK_DL(uv_dlsym(&zlib, "inflate", (void **) &inflate_f));
+    CHECK_DL(uv_dlsym(&zlib, "zError", (void **) &zError_f));
+#endif
+
+    ZLibVersion = zlib_ver();
+    if (ZLibVersion[0] != ZLIB_VERSION[0]) {
+        return;
+    }
+    if (zlib_flags() & NO_GZIP) {
+        encodings = "deflate";
+    } else {
+        encodings = "gzip, deflate";
+    }
+    goto done;
+
+    on_error:
+    UM_LOG(ERR, "failed to initialize HTTP decompression: %s", uv_dlerror(&zlib));
+
+    done:
+    return;
+}
+
+const char *um_available_encoding() {
+    uv_once(&init_guard, init);
+    return encodings;
+}
+
+#define inflateInit_ inflateInit_f
+#define inflateInit2_ inflateInit2_f
+
+http_inflater_t *um_get_inflater(const char *encoding, data_cb cb, void *ctx) {
+    um_available_encoding();
+
+    http_inflater_t *inf = calloc(1, sizeof(http_inflater_t));
+    inf->s.zalloc = comp_alloc;
+    inf->s.zfree = comp_free;
+    if (strcmp(encoding, "gzip") == 0)
+        inflateInit2(&inf->s, 16 + MAX_WBITS);
+    else if (strcmp(encoding, "deflate") == 0)
+        inflateInit(&inf->s);
+    else {
+        free(inf);
+        return NULL;
+    }
+
+    inf->cb = cb;
+    inf->cb_ctx = ctx;
+    return inf;
+}
+
+void um_free_inflater(http_inflater_t *inflater) {
+    if (inflater) {
+        inflateEnd_f(&inflater->s);
+        free(inflater);
+    }
+}
+
+int um_inflate(http_inflater_t *inflater, const char *compressed, size_t len) {
+    inflater->s.next_in = compressed;
+    inflater->s.avail_in = len;
+    char decompressed[32 * 1024];
+    while(inflater->s.avail_in > 0) {
+        inflater->s.next_out = decompressed;
+        inflater->s.avail_out = sizeof(decompressed);
+        int rc = inflate_f(&inflater->s, Z_FULL_FLUSH);
+        ssize_t decomp_count = sizeof(decompressed) - inflater->s.avail_out;
+        if (decomp_count > 0) {
+            inflater->cb(inflater->cb_ctx, decompressed, decomp_count);
+        }
+        if (rc == Z_STREAM_END) {
+            inflater->complete = 1;
+            break;
+        }
+    }
+    return 0;
+}
+
+int um_inflate_state(http_inflater_t *inflater) {
+    if (inflater->s.msg) return -1;
```

Comment:
you have tests checking for um_inflate_state = 1 and 0 but none for when it's -1. is it worthwhile to test for this condition?